src/Engine/ProtoCore/DSASM/Heap.cs

```diff
@@ -687,7 +717,10 @@ private int RecursiveMark(StackValue root)
         /// </summary>
         private void StartCollection()
         {
-            sweepSet = new HashSet<int>(Enumerable.Range(0, heapElements.Count));
+            // We start from a clean sweepSet
+            sweepSet.Clear();
+
+            sweepSet.UnionWith(Enumerable.Range(0, heapElements.Count));
```

Comment:
UnionWith seems to work fine for this (concat does not work for some reason)

Reply:
Is this simply pre-allocating space in the hash set?

Reply:
no, it is actually adding all indexes from 0 to HeapElements.Count
basically takes a snapshot of the heap (using indexes)

Reply:
I think `concat` may not work because it returns another IEnumerable - which means that it may use deferred execution - so the snapshot is not taken when you intended it.

With `UnionWith`, this returns a Set and should execute the `Enumerable.Range` call.

Reply:
Sounds correct